ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/SelectionUtil.java

```diff
@@ -65,13 +71,27 @@ public static LogicalSchema buildProjectionSchema(
           && parentSchema.isKeyColumn(((ColumnReferenceExp) expression).getColumnName());
 
       if (keyColumn) {
-        builder.keyColumn(select.getAlias(), expressionType);
-        keys.put(select.getAlias(), expressionType);
+        sourceKeyToAliasAndType.put(
+            ((ColumnReferenceExp) expression).getColumnName(),
+            new Pair<>(select.getAlias(), expressionType)
+        );
       } else {
         builder.valueColumn(select.getAlias(), expressionType);
       }
     }
 
+    // keys need to be added in the same order as the source
+    for (final Column key : parentSchema.key()) {
+      final List<Pair<ColumnName, SqlType>> aliasAndTypes = sourceKeyToAliasAndType.get(key.name());
+      for (final Pair<ColumnName, SqlType> aliasAndType : aliasAndTypes) {
+        if (aliasAndType != null) {
+          // can be null if the key was not selected - this will cause a failure
+          // down the line but here we just ignore it here
```

Comment:
Why can it case an error? And how would it surface to the user? -- Not sure what we don't want to catch this error right here?

Reply:
you're required to select all keys, if you don't it'll complain :) I don't think we should add unnecessary redundant checks in places they don't belong. there's no reason why this specific code needs to know that you're required to select all keys. The code that requires that should enforce it.